feat(ui): add chonk UI feature flag to theme

TkDodo opened the pull request:
This PR adds an `isChonk:boolean` flag to the `theme`, which will be set to true depending on the organization feature flag `'chonk-ui'`.

This extra flag will allow us to move fast when having to change one-off styled-components, because we can now access that flag from the theme, for example:

```
color: ${p => p.theme.isChonk ? ... : ... };
```

Note: toggling the feature flag in the dev-toolbar does _not_ immediately update the `organization` in the `OrganizationStore`, so a full page reload is needed for this to take effect.

TkDodo committed:
feat(ui): add chonk UI feature flag to theme

TkDodo committed:
test: add a test for isChonk flag on the theme

TkDodo committed:
Merge branch 'master' into tkdodo/feat/is-chonk

TkDodo commented on static/app/views/app/index.tsx:
```diff
@@ -270,25 +272,40 @@ function App({children, params}: Props) {
-                </GlobalDrawer>
-              </GlobalFeedbackForm>
-            </AsyncSDKIntegrationContextProvider>
+            <ChonkThemeProvider>
```
note: this is the top-most place where we can have access to the `OrganizationProvider`. There’s a `ThemeAndStyleProvider` at the very root of the application, but it can’t read the organization from `useOrganization`.

Another idea would’ve been to move `ThemeAndStyleProvider` down here to be under the `OrganizationProvider`, but given that the `chonk-ui` feature flag is temporary, I think adding another theme provider is the better way.

TkDodo commented on static/app/views/app/index.tsx:
```diff
@@ -270,25 +272,40 @@ function App({children, params}: Props) {
 }
 
+function ChonkThemeProvider({children}: {children: React.ReactNode}) {
+  const organization = useOrganization({allowNull: true});
```
There’s one render cycle where `organization` is still `null` when this renders. This makes sense because the `OrganizationProvider` doesn’t itself load the organization.

We could alternatively move the `ChonkThemeProvider` further down the tree - e.g. into the `OrganizationLayout` below the `OrganizationContainer`. There, accessing the flag would be safe. But I think since we have a boolean flag, working with a falsy value is also fine.

TkDodo committed:
fix: tests

setNoOrganization works when only one test is run, but hangs indefinitely when multiple tests are run...

TkDodo committed:
Merge branch 'master' into tkdodo/feat/is-chonk

natemoo-re commented on static/app/views/app/index.tsx:
```diff
@@ -270,25 +272,40 @@ function App({children, params}: Props) {
   );
 }
 
+function ChonkThemeProvider({children}: {children: React.ReactNode}) {
```
Comment for clarity?
```suggestion
/** 
  * Temporary provider for new UI2 theme rollout
  */
function ChonkThemeProvider({children}: {children: React.ReactNode}) {
```

natemoo-re commented on static/app/views/app/index.tsx:
```diff
@@ -270,25 +272,40 @@ function App({children, params}: Props) {
 
+function ChonkThemeProvider({children}: {children: React.ReactNode}) {
+  const organization = useOrganization({allowNull: true});
+  const theme = useTheme();
+
+  const chonkTheme = useMemo(() => ({...theme, isChonk: true}), [theme]);
```
Looks like we can skip the `useTheme()` and `useMemo()` calls. According to the docs, passing `{ isChonk: true }` here would be merged with the ancestor theme.

> The theme object will be merged into its ancestor as if by [Object.assign](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign)
> - [`ThemeProvider` docs](https://emotion.sh/docs/theming#themeprovider-reactcomponenttype)

natemoo-re replied:
`theme` also accepts a callback, so if we want this to be the compatability layer which modifies the original theme, that would work too.

TkDodo replied:
good to know!

> The theme object will be merged into its ancestor as if by [Object.assign](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign)

I don’t think we should do that, as it would permanently overwrite our constant `lightTheme` / `darkTheme`. If we re-use those in other places, we would get the modified version.

Okay I looked up the source code and they don’t use `Object.assign`, but create a new object instead:

https://github.com/emotion-js/emotion/blob/cce67ec6b2fc94261028b4f4778aae8c3d6c5fd6/packages/react/src/theming.tsx#L57

so it should be fine. But also, the theme is always re-created with an identity check:

https://github.com/emotion-js/emotion/blob/cce67ec6b2fc94261028b4f4778aae8c3d6c5fd6/packages/react/src/theming.tsx#L74-L76

so hoisting it outside of the component seems best:

62523700

natemoo-re reviewed (approved):
Nice test + seems like a reasonable tradeoff! Seems like we can simplify the `theme` value a bit before merging, but LGTM

TkDodo committed:
Update static/app/views/app/index.tsx

Co-authored-by: Nate Moore <natemoo-re@users.noreply.github.com>

getsantry[bot] committed:
:hammer_and_wrench: apply pre-commit fixes

TkDodo committed:
ref: pass partial theme to ThemeProvider

this leverages the ThemeProvider's functionality of accepting a partial theme, as they get merged together internally; The theme should be referentially stable though to avoid having this merge happen on every render.

TkDodo committed:
Merge branch 'master' into tkdodo/feat/is-chonk

TkDodo committed:
Merge branch 'master' into tkdodo/feat/is-chonk
